crypto/rand/rand_lib.c

```diff
@@ -854,6 +854,9 @@ int rand_priv_bytes_ex(OPENSSL_CTX *ctx, unsigned char *buf, int num)
     int ret;
     const RAND_METHOD *meth = RAND_get_rand_method();
 
+    if (meth == NULL)
+        return 0;
+
```

Comment:
@paulidale I see that a NULL pointer is returned in FIPS mode on purpose. Does this mean that this code path here is never intended to be reached in FIPS mode, or am I missing something?

https://github.com/openssl/openssl/blob/04bb55daeafc6da36f4445c26cc95543e9c04e51/crypto/rand/rand_lib.c#L770-L774

Reply:
Or maybe @mattcaswell is the right person to ask: What was the reason why you added the NULL return in FIPS mode? The commit message below does not give an indication about it.

commit a2f27fd750b9ae62a571a9212c7154889100bdb0
Author: Matt Caswell <matt@openssl.org>
Date:   Fri May 24 16:36:44 2019 +0100

    Move the rand_nonce_lock code into drbg_lib.c
    
    It was previously rand_lib but it makes more sense in drbg_lib.c since
    all the functions that use this lock are only ever called from drbg_lib.c
    
    We add some FIPS_MODE defines in preparation for later moving this code
    into the FIPS module.
    
    Reviewed-by: Richard Levitte <levitte@openssl.org>
    (Merged from https://github.com/openssl/openssl/pull/9039)

Reply:
Note: It seems to me that the FIPS changes are not final yet. At least, they are not consistent:

While for `RAND_get_rand_method()` only the implementation is removed in FIPS mode,

https://github.com/openssl/openssl/blob/04bb55daeafc6da36f4445c26cc95543e9c04e51/crypto/rand/rand_lib.c#L770-L774

the `RAND_set_rand_engine()` method is removed entirely,


https://github.com/openssl/openssl/blob/04bb55daeafc6da36f4445c26cc95543e9c04e51/crypto/rand/rand_lib.c#L804-L807

even though there seems to be no FIPS-replacement for this public API function (yet).

Reply:
The FIPS RAND stuff is in flux.  Moving the DRBG/seeding into a provider is proving troublesome.  The code needs to be split both along the FIPS/non-FIPS axis but also along an algorithm/context one.  I'm also trying to merge the DRBG and seeding into a single algorithm class.  Maybe I'm being too ambitious :grin:

Reply:
> Maybe I'm being too ambitious :grin:

I know that feeling...

Reply:
> The FIPS RAND stuff is in flux. 

Ok, I'll just ignore the current fluctuations and temporary inconsistencies and only add null pointer checks where necessary.

Reply:
maybe going back to just one supported algorithm (aes-ctr) would help simplify?

Reply:
The actual DRBGs are looking like being the easy bit.